src/object/obj_tx.c

```diff
@@ -640,154 +673,774 @@ dc_tx_open(tse_task_t *task)
 }
 
 struct tx_commit_cb_args {
-	struct dc_tx	*tcca_tx;
-	crt_rpc_t	*tcca_req;
+	struct dc_tx		*tcca_tx;
+	crt_rpc_t		*tcca_req;
+	daos_tx_commit_t	*tcca_args;
 };
 
 static int
 dc_tx_commit_cb(tse_task_t *task, void *data)
 {
-	struct tx_commit_cb_args *tcca = (struct tx_commit_cb_args *)data;
+	struct tx_commit_cb_args *tcca = data;
 	struct dc_tx		 *tx = tcca->tcca_tx;
 	crt_rpc_t		 *req = tcca->tcca_req;
+	struct obj_cpd_out	 *oco = crt_reply_get(req);
+	uint64_t		 *sub_epochs;
 	int			  rc = task->dt_result;
+	bool			  locked = true;
 
 	D_MUTEX_LOCK(&tx->tx_lock);
 
+	if (rc == 0) {
+		rc = oco->oco_ret;
+		if (rc == 0) {
+			int	*sub_rets = oco->oco_sub_rets.ca_arrays;
+
+			/* FIXME: currently, we pack one DTX in one CPD RPC. */
+			rc = sub_rets[0];
+		}
+	}
+
 	if (rc == 0) {
 		tx->tx_status = TX_COMMITTED;
+		tx->tx_retry = 0;
+		dc_tx_cleanup(tx);
+
+		sub_epochs = oco->oco_sub_epochs.ca_arrays;
+		if (tx->tx_epoch.oe_value == 0)
+			/* FIXME: currently, we pack one DTX in one CPD RPC. */
+			tx->tx_epoch.oe_value = sub_epochs[0];
+
+		if (tx->tx_epoch.oe_value == 0)
+			D_WARN("Server forgot to reply epoch for TX "DF_DTI"\n",
+			       DP_DTI(&tx->tx_id));
+
 		goto out;
 	}
 
 	tx->tx_status = TX_FAILED;
 
-	/* Required to restart the TX with newer epoch. */
-	if (rc == -DER_TX_RESTART)
+	/* Be required to restart the TX with newer epoch. */
+	if (rc == -DER_TX_RESTART) {
+		tx->tx_pm_ver = oco->oco_map_version;
+		task->dt_result = rc;
+		/* Set retry flag even if it is restarted. */
+		tx->tx_retry = 1;
+
+		sub_epochs = oco->oco_sub_epochs.ca_arrays;
+		/* FIXME: currently, we pack one DTX in one CPD RPC. */
+		if (tx->tx_epoch.oe_value < sub_epochs[0])
+			tx->tx_epoch.oe_value = sub_epochs[0];
+
 		goto out;
+	}
+
+	D_ASSERTF(tx->tx_pm_ver >= oco->oco_map_version,
+		  "Invalid pm version for TX "DF_DTI": TX ver %u, RPC ver %u\n",
+		  DP_DTI(&tx->tx_id), tx->tx_pm_ver, oco->oco_map_version);
+
+	if (!obj_retry_error(rc)) {
+		task->dt_result = rc;
+		tx->tx_retry = 0;
+
+		goto out;
+	}
 
-	/* FIXME:
-	 * 1. Check pool map, if pool map is refreshed, restart TX.
-	 * 2. Retry the RPC if it is retriable error.
-	 * 3. Failout for other failures.
+	/* Keep the retry flag until TX is closed. */
+	tx->tx_retry = 1;
+	tcca->tcca_args->flags |= DTF_RETRY_COMMIT;
+
+	D_MUTEX_UNLOCK(&tx->tx_lock);
+	locked = false;
```

Comment:
The lock contention on tx_lock isn't really important when we're committing the TX, no? Dropping tx_lock below seems just fine, unless numbers prove otherwise.

Reply:
That depends on how the subsequent dc_task_resched() that may take more other locks and may re-entry DTX logic, all depends on such function implementation. Since we do not need such lock any longer, release it earlier may be more safe.

Reply:
No strong opinions.